src/pb_stub.cc

```diff
@@ -793,26 +793,39 @@ Stub::ProcessRequests(RequestBatch* request_batch_shm_ptr)
           std::to_string(response_size) + "\n";
       throw PythonBackendException(err);
     }
-    for (auto& response : responses) {
+
+    for (size_t i = 0; i < response_size; i++) {
       // Check the return type of execute function.
-      if (!py::isinstance<InferResponse>(response)) {
-        std::string str = py::str(response.get_type());
-        throw PythonBackendException(
-            std::string("Expected an 'InferenceResponse' object in the execute "
-                        "function return list, found type '") +
-            str + "'.");
+      InferRequest* infer_request = py_request_list[i].cast<InferRequest*>();
+      if (infer_request->ReleaseFlags() ==
+          TRITONSERVER_REQUEST_RELEASE_RESCHEDULE) {
```

Comment:
I don't think the Python backend should enforce how to send the response, here it enforces the response must be sent with a request release ALL flag.

There is legacy implementation that Python backend have restriction on the response behavior depending on whether the model is marked decoupled. But in terms of the backend API usage, it is completely fine for the user model to send one response on reschedule and send response end flag alone on executing the rescheduled request (at the end one and only one response is sent).

Reply:
For non-decoupled model, there is only one response per request. I think there are two possible cases:

1.  **First iteration**: request reschedule flag + no response returned(append `None` to the return list of responses)
**Final iteration**: response complete flag + response
2. **First iteration**: rescheduled flag + response
**Second iteration(optional)**: reschedule flag + no response returned
**Final iteration**: response complete flag + no response returned

The above cases should be valid for backend API usage. However, for Python backend, the `response_sender` is only exposed to users for decoupled model, which means currently it's not possible for case 2 - sending response and response complete flag in different iterations. I think it makes sense to support case 2, and we can modify the Python backend to also have a monitor thread for non-decoupled models so that case 2 can be satisfied.
@Tabrizian Will there be any issues if we expose `response_sender` to non-decoupled models to fulfill the use case?
cc @nnshah1

Reply:
@krishung5 could you please elaborate why the 1-1 mapping is limiting here? What is the problem that we're trying to solve. If the user wants to send less than one response, they can use decoupled mode.

Reply:
Resolved offline.